### PR TITLE
Add DecimateRecording

### PR DESCRIPTION
**Pull request by TomBugnon**
Hi,

I needed this because I want to downsample while keeping track of the actual underlying frame ids. In my case I don't care about aliasing etc because I will apply it on a smoothed signal, but it might be a good idea to add some safeguards... Maybe changing the name to avoid confusion with the scipy.decimate function?

Best
Tom

### CODE REVIEW

File: src/spikeinterface/preprocessing/decimate.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    Important: This uses simple array slicing for decimation rather than eg scipy.decimate.
+    This might introduce aliasing, or skip across signal of interest.
+    Consider  spikeinterface.preprocessing.ResampleRecording for safe resampling.
+
+    Parameters
+    ----------
+    recording : Recording
+        The recording extractor to be re-referenced
+    decimation_frame_step : int
+    decimation_frame_start : int, default 0
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:24]: Can you add descriptions?

Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:24]:
```suggestion
    decimation_frame_step : int
    decimation_frame_start : int, default: 0
```
We use `default: ...`  for default values

File: src/spikeinterface/preprocessing/tests/test_decimate.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+def test_decimate():
+    rec = generate_recording()
+
+    N = 101
+    rec = NumpyRecording([np.arange(N).reshape(N,1)], 1)
+    parent_traces = rec.get_traces()
+
+    for decimation_frame_start, decimation_frame_step in itertools.product(
+        [0, 1, 9, 10, 11, 200],
+        [1, 9, 10, 11, 200]
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/tests/test_decimate.py:20]: It would be great to use `@pytest.mark.parametrize` to expand tests, see for example [here](https://github.com/SpikeInterface/spikeinterface/blob/main/src/spikeinterface/preprocessing/tests/test_interpolate_bad_channels.py#L84-L88)

**Review by alejoe91 — CHANGES_REQUESTED** [review]
Thanks @TomBugnon 

A few small comments!

File: src/spikeinterface/preprocessing/tests/test_decimate.py (hunk 2 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/tests/test_decimate.py]: got lazy sorry

**Comment by alejoe91** [comment]
Do you mind also running pre-commit (or just black) locally? Since the PR is from an organization, the bot cannot push to it

**Comment by samuelgarcia** [comment]
Super. Merci Tom.
Did you test corner case like get_trace on borders with one sample at the end ?

File: src/spikeinterface/preprocessing/decimate.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        )
+        self._parent_segment = parent_recording_segment
+        self._decimation_frame_step = decimation_frame_step
+        self._decimation_frame_start = decimation_frame_start
+        self._dtype = dtype
+
+    def get_num_samples(self):
+        return len(
+            range(self._decimation_frame_start, self._parent_segment.get_num_samples(), self._decimation_frame_step)
+        )
```
Comment (samuelgarcia) [inline src/spikeinterface/preprocessing/decimate.py:111]: This is smart but maybe something more explicit with rounding or ceiling integer could be more usefull for debugging no ?

File: src/spikeinterface/preprocessing/decimate.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    def get_traces(self, start_frame, end_frame, channel_indices):
+        if start_frame is None:
+            start_frame = 0
+        if end_frame is None:
+            end_frame = self.get_num_samples()
+
+        # Account for offset and end when querying parent traces
+        parent_start_frame = self._decimation_frame_start + start_frame * self._decimation_frame_step
+        parent_end_frame = parent_start_frame + (end_frame - start_frame) * self._decimation_frame_step
```
Comment (samuelgarcia) [inline src/spikeinterface/preprocessing/decimate.py:121]: How can we be sure that this is not longer than the parent size ?
this is also related to the get_nump_samples() to be sure of ceiling vs rouding.

File: src/spikeinterface/preprocessing/decimate.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        # Original sampling frequency
+        self._orig_samp_freq = recording.get_sampling_frequency()
+        if not isinstance(decimation_frame_step, int) or decimation_frame_step <= 0:
+            raise ValueError(f"Expecting strictly positive integer for `decimation_frame_step` arg")
+        self._decimation_frame_step = decimation_frame_step
+        if not isinstance(decimation_frame_start, int) or decimation_frame_step < 0:
+            raise ValueError(f"Expecting positive integer for `decimation_frame_step` arg")
+        self._decimation_frame_start = decimation_frame_start
+        resample_rate = self._orig_samp_freq / self._decimation_frame_step
+        self._sampling_frequency = resample_rate
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:54]:
```suggestion
```

File: src/spikeinterface/preprocessing/decimate.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        # fix_dtype not always returns the str, make sure it does
+        dtype = fix_dtype(recording, None).str
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:56]: You can remove this too, since it's inherited from the parent by default in the `BasePreprocessor`
```suggestion
```

File: src/spikeinterface/preprocessing/decimate.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+        BasePreprocessor.__init__(self, recording, sampling_frequency=resample_rate, dtype=dtype)
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:58]:
```suggestion
        BasePreprocessor.__init__(self, recording, sampling_frequency=resample_rate, dtype=None)
```

File: src/spikeinterface/preprocessing/tests/test_decimate.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+import itertools
+from spikeinterface import NumpyRecording
+from spikeinterface.core import generate_recording
+from spikeinterface.preprocessing.decimate import DecimateRecording
+import numpy as np
+
+
+@pytest.mark.parametrize("decimation_frame_start", [0, 1, 9, 10, 11, 200])
+@pytest.mark.parametrize("decimation_frame_step", [1, 9, 10, 11, 200])
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/tests/test_decimate.py:12]: A couple of comments:

- shouldn't `decimation_frame_start` < `decimation_frame_step` be enforced?
- shouldn't we assert that `decimation_frame_start`  < `num_samples`?

File: src/spikeinterface/preprocessing/decimate.py (hunk 4 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py:121]: It can be longer than the parent size if `end_frame > self.get_num_samples()`, but that will still return the correct output I think. 

This is related to the discussion about enforcing `end_frame <= num_samples` some time ago, not sure what was decided then?

File: src/spikeinterface/preprocessing/decimate.py (hunk 3 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py:111]: ok ok no laziness allowed that's fine

File: src/spikeinterface/preprocessing/decimate.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    def __init__(
+        self,
+        recording,
+        decimation_factor,
+        decimation_offset=0,
+    ):
+        # Original sampling frequency
+        self._orig_samp_freq = recording.get_sampling_frequency()
+        parent_nsamp = recording.get_num_samples()
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/decimate.py:50]:
```suggestion
        parent_nsamp = min([recording.get_num_samples(segment_index) for segment_index in range(recording.get_num_segments()])
```
Parent can be multi-segment! ;)

File: src/spikeinterface/preprocessing/decimate.py (hunk 4 above)
Comment (alejoe91) [reply src/spikeinterface/preprocessing/decimate.py]: yes, so the majority vote was to be strict about this.

I think it is ok in this case to have end_frame larger than the parent num samples (it shouldn't throuw an error), but capping at parent segment num samples won't hurt

File: src/spikeinterface/preprocessing/tests/test_decimate.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+from pathlib import Path
+
+import itertools
+from spikeinterface import NumpyRecording
+from spikeinterface.core import generate_recording
+from spikeinterface.preprocessing.decimate import DecimateRecording
+import numpy as np
+
+
+@pytest.mark.parametrize("decimation_offset", [0, 1, 9, 10, 11, 100, 101])
```
Comment (alejoe91) [inline src/spikeinterface/preprocessing/tests/test_decimate.py:11]: Thanks! On the same line, shouldn't `decimation_offset=101` fail?

In #1989 we (kinda) agreed that we shouldn't allow for empty arrays returns. So if you have 101 samples and you ask for `parent_traces[101::*]` it will return an empty array.

File: src/spikeinterface/preprocessing/decimate.py (hunk 4 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py]: I clipped start_frame/end_frame to num_samp.
Current behavior does not seem to be standardized in other preprocessors so I'm not sure exactly what to enforce here... Should I check that start_frame/end_frame are positive integers?

File: src/spikeinterface/preprocessing/tests/test_decimate.py (hunk 10 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/tests/test_decimate.py:11]: Not sure what you mean... I check that some parameter combinations raise ValueErrors [here](https://github.com/SpikeInterface/spikeinterface/blob/6b07f9423d7954fb3d49c6f4b4b859c2e3c031de/src/spikeinterface/preprocessing/tests/test_decimate.py#L22) (couldn't find an elegant way to do it with `parameterize`

File: src/spikeinterface/preprocessing/decimate.py (hunk 9 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py:50]: OK so should what should we do for multi-segments...
Same decimation factor for all segments, and accept decimation_offset to be either an int or an array of ints ?

Comment (alejoe91) [reply src/spikeinterface/preprocessing/decimate.py:50]: I think same for both is fine

File: src/spikeinterface/preprocessing/tests/test_decimate.py (hunk 10 above)
Comment (alejoe91) [reply src/spikeinterface/preprocessing/tests/test_decimate.py:11]: totally missed that! Looks great!

File: src/spikeinterface/preprocessing/decimate.py (hunk 9 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py]: ok done

File: src/spikeinterface/preprocessing/decimate.py (hunk 4 above)
Comment (TomBugnon) [reply src/spikeinterface/preprocessing/decimate.py]: This is the last point (hopefully)

**Review by alejoe91 — APPROVED** [review]
